test/extended/prometheus/prometheus.go

```diff
@@ -42,6 +44,225 @@ import (
+						severity, severityRe.String(),
+					),
+				)
+			}
+
+			return nil
+		})
+	})
+
+	g.It("should have description and summary annotations", func() {
+		forEachAlertingRule(alertingRules, func(alert alertingRule) sets.String {
+			if descriptionExceptions.Has(alert.Name) {
+				framework.Logf("Alerting rule %q is known to have missing annotations.")
+				return nil
+			}
+
+			violations := sets.NewString()
+
+			if _, found := alert.Annotations["description"]; !found {
+				// If there's no 'description' annotation, but there is a
+				// 'message' annotation, suggest renaming it.
+				if _, found := alert.Annotations["message"]; found {
+					violations.Insert("has no 'description' annotation, but has a 'message' annotation." +
+						" OpenShift alerts must use 'description' -- consider renaming the annotation")
+				} else {
+					violations.Insert("has no 'description' annotation")
+				}
+			}
+
+			if _, found := alert.Annotations["summary"]; !found {
+				violations.Insert("has no 'summary' annotation")
+			}
+
+			return violations
+		})
+	})
+
+	g.It("should have a runbook_url annotation if the alert is critical", func() {
+		violations := sets.NewString()
+
+		forEachAlertingRule(alertingRules, func(alert alertingRule) sets.String {
+			severity := alert.Labels["severity"]
+			runbook := alert.Annotations["runbook_url"]
+
+			if severity == "critical" && runbook == "" {
+				violations.Insert(
+					fmt.Sprintf("WARNING: Alert %q is critical and has no 'runbook_url' annotation", alert.Name),
+				)
+			} else if runbook != "" {
+				// If there's a 'runbook_url' annotation, make sure it's a
+				// valid URL and that we can fetch the contents.
+				if err := validateURL(runbook, 10*time.Second); err != nil {
+					violations.Insert(
+						fmt.Sprintf("WARNING: Alert %q has an invalid 'runbook_url' annotation: %v",
+							alert.Name, err),
+					)
+				}
+			}
+
+			return nil // Always return nil, because these aren't required yet.
+		})
+
+		// These aren't enforced yet.  They are just reporting the errors by
+		// calling the g.Skip() method.
+		if len(violations) > 0 {
+			g.Skip(
+				fmt.Sprintf("Incompliant rules detected:\n\n%s", strings.Join(violations.List(), "\n")),
+			)
+		}
+	})
+})
+
+// RuleDiscovery has info for all rules
+type ruleDiscovery struct {
+	RuleGroups []*ruleGroup `json:"groups"`
+}
+
+// RuleGroup has info for rules which are part of a group
+type ruleGroup struct {
+	Name string `json:"name"`
+	File string `json:"file"`
+	// Rules contain alertingRule and recordingRule structs.
+	Rules          []json.RawMessage `json:"rules"`
+	Interval       float64           `json:"interval"`
+	EvaluationTime float64           `json:"evaluationTime"`
+	LastEvaluation time.Time         `json:"lastEvaluation"`
+}
+
+const (
+	alertingRuleType  = "alerting"
+	recordingRuleType = "recording"
+)
+
+type ruleType struct {
+	Type string `json:"type"`
+}
+
+type alertingRule struct {
```

Comment:
Couldn't we import these types from the `client_golang` library? https://pkg.go.dev/github.com/prometheus/client_golang/api/prometheus/v1

Reply:
Honestly, that was left over from the branch I based this on. Not sure why it was like that. Fixed.